raphtory-graphql/src/data.rs

```
@@ -1,68 +1,90 @@
 use raphtory::{
     core::Prop,
     db::api::view::MaterializedGraph,
     prelude::{GraphViewOps, PropUnwrap, PropertyAdditionOps},
     search::IndexedGraph,
-    vectors::vectorised_graph::DynamicVectorisedGraph,
 };
-use std::{collections::HashMap, fs, path::Path, sync::Arc};
+use std::{
+    collections::HashMap,
+    fs,
+    path::{Path, PathBuf},
+    sync::Arc,
+};
 use walkdir::WalkDir;
 
-#[derive(Default)]
 pub struct Data {
-    pub(crate) graphs: Arc<RwLock<HashMap<String, IndexedGraph<MaterializedGraph>>>>,
-    pub(crate) vector_stores: Arc<RwLock<HashMap<String, DynamicVectorisedGraph>>>,
+    pub(crate) work_dir: String,
+    pub(crate) graphs: Arc<Cache<String, IndexedGraph<MaterializedGraph>>>,
+    pub(crate) global_plugins: GlobalPlugins,
 }
 
 impl Data {
-    pub fn from_map<G: Into<MaterializedGraph>>(graphs: HashMap<String, G>) -> Self {
-        let graphs = Arc::new(RwLock::new(Self::convert_graphs(graphs)));
-        let vector_stores = Arc::new(RwLock::new(HashMap::new()));
-        Self {
-            graphs,
-            vector_stores,
-        }
-    }
+    pub fn new(
+        work_dir: &Path,
+        maybe_graphs: Option<HashMap<String, MaterializedGraph>>,
+        maybe_graph_paths: Option<Vec<PathBuf>>,
+        maybe_cache_config: Option<CacheConfig>,
+    ) -> Self {
+        let cache_configs = if maybe_cache_config.is_some() {
+            maybe_cache_config.unwrap()
+        } else {
+            let app_config = load_config().expect("Failed to load config file");
+            app_config.cache
+        };
+
+        let graphs_cache_builder = Cache::builder()
+            .max_capacity(cache_configs.capacity)
+            .time_to_idle(std::time::Duration::from_secs(cache_configs.tti_seconds))
+            .build();
+
+        let graphs_cache: Arc<Cache<String, IndexedGraph<MaterializedGraph>>> =
+            Arc::new(graphs_cache_builder);
+
+        save_graphs_to_work_dir(work_dir, &maybe_graphs.unwrap_or_default())
```

As we discussed - this should take an overwrite argument - of which if it is false we need to throw an error for the user saying that the file already exists - this should be checked first so we dont end up in the case where half the graphs are savedd.